pkg/controller/apiextensions/composite/composed/composed.go

```diff
@@ -0,0 +1,181 @@
+func WithConnectionDetailFetcher(cdf ConnectionDetailsFetcher) ComposerOption {
+	return func(composer *Composer) {
+		composer.ConnectionDetailsFetcher = cdf
+	}
+}
+
+// WithOverlayApplicator returns a ComposerOption that changes the
+// OverlayApplicator of Composer.
+func WithOverlayApplicator(oa OverlayApplicator) ComposerOption {
+	return func(composer *Composer) {
+		composer.OverlayApplicator = oa
+	}
+}
+
+// WithConfigurator returns a ComposerOption that changes the Configurator of
+// Composer.
+func WithConfigurator(c Configurator) ComposerOption {
+	return func(composer *Composer) {
+		composer.Configurator = c
+	}
+}
+
+// ComposerOption configures the Composer object.
+type ComposerOption func(*Composer)
+
+// NewComposer returns a new Composer that composes infrastructure resources
+// in a Kubernetes API server.
+func NewComposer(kube client.Client, opts ...ComposerOption) *Composer {
+	c := &Composer{
+		client: resource.ClientApplicator{
+			Client:     kube,
+			Applicator: resource.NewAPIPatchingApplicator(kube),
+		},
+		Configurator:             &DefaultConfigurator{},
+		OverlayApplicator:        &DefaultOverlayApplicator{},
+		ConnectionDetailsFetcher: &APIConnectionDetailsFetcher{client: kube},
+	}
+
+	for _, f := range opts {
+		f(c)
+	}
+	return c
+}
+
+// An Composer composes infrastructure resources in a Kubernetes API server.
+type Composer struct {
+	client resource.ClientApplicator
+	Configurator
+	OverlayApplicator
+	ConnectionDetailsFetcher
+}
+
+// Compose the supplied Composed resource into the supplied Composite resource
+// using the supplied CompositeTemplate.
+func (r *Composer) Compose(ctx context.Context, cp resource.Composite, cd resource.Composed, t v1alpha1.ComposedTemplate) (Observation, error) {
+
+	// Composed may not have a name if it's not created yet.
+	if cd.GetName() != "" {
+		nn := types.NamespacedName{Namespace: cd.GetNamespace(), Name: cd.GetName()}
+		if err := r.client.Get(ctx, nn, cd); resource.IgnoreNotFound(err) != nil {
+			return Observation{}, errors.Wrap(err, errGetComposed)
+		}
+	}
+
+	if !meta.WasCreated(cd) {
+		// Configure is called only if Composed resource is not created yet
+		// because the fields that are not referred in the patches should be
+		// mutable by the user, i.e. dependency on the Composition should be
+		// limited to the field propagation.
+		if err := r.Configure(cp, cd, t); err != nil {
+			return Observation{}, errors.Wrap(err, errConfigure)
+		}
+	}
+
+	// Overlay is applied to the Composed resource in all cases so that we can
+	// keep Composed resource up-to-date with the changes in Composite resource.
+	if err := r.Overlay(cp, cd, t); err != nil {
+		return Observation{}, errors.Wrap(err, errOverlay)
+	}
+
+	// Connection details are fetched in all cases in a best-effort mode, i.e.
+	// it doesn't return error if the secret does not exist or the resource
+	// does not publish a secret at all.
+	conn, err := r.Fetch(ctx, cd, t)
```

Comment:
I think this would read better if it weren't an embedded interface. It's not obvious here what is being fetched.

Reply:
Yeah, thanks for the suggestion!